src/ai/AIConstruction.cpp

```diff
@@ -593,6 +593,16 @@ bool AIConstruction::MilitaryBuildingSitesLimit()
     return complete + 3 > inconstruction;
 }
 
+bool AIConstruction::IncrementWanted(BuildingType type)
+{
+    if (GetBuildingCount(type) <= buildingsWanted[type])
```

Comment:
Shouldn't this be >=? 
And I think putting the increment and the check in the same function is misleading for the reader of the call. It is also not clear why this leads to forceBuild so checking the count there would be better. So this function can actually be removed

Reply:
That check's there so that AI doesn't keep building and building new military buildings if first one isn't complete yet so the "<=" is correct. I guess the function can be removed and check moved to the AIJHHelper.cpp, but I thought it would be more proper to actually update the wanted list.

Reply:
But why would you want to increase the wanted buildings if there are more wanted than built? 
And this implementation won't avoid the case you are talking about: built=0 wanted=1 so increase wanted to 2. Then to 3 etc. 
Maybe rather a) build military bld even if no more wanted in case of space problems and b) don't do a) of there are military blds in construction.
Maybe even make sure we don't add military buildings to the wanted buildings unless we have a sawmill (at least as a building site) or can't build it (so add it to wanted buildings if no position is found, similar to your solution)